### PR TITLE
ttnctl gateways

### PR DESCRIPTION
**Pull request by romeovs**
Adds additional commands to `ttnctl gateways`:

```
Available Commands:
  delete      Delete a gateway
  edit        edit a gateway
  list        List your gateways
  register    Register a gateway
  status      Get status of a gateway
```

### CODE REVIEW

**Comment by htdvisser** [comment]
Please add `Example`s to commands (see other `ttnctl` commands)

File: ttnctl/cmd/gateways_delete.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"github.com/TheThingsNetwork/ttn/api"
+	"github.com/TheThingsNetwork/ttn/ttnctl/util"
+	"github.com/spf13/cobra"
+)
+
+var gatewaysDeleteCmd = &cobra.Command{
+	Use:   "delete [GatewayID]",
+	Short: "Delete a gateway",
+	Long:  `ttnctl gateways deletw can be used to delete a gateway`,
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_delete.go]: `deletw` typo

File: ttnctl/cmd/gateways_delete.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+			cmd.UsageFunc()(cmd)
+			return
+		}
+
+		gatewayID := args[0]
+		if !api.ValidID(gatewayID) {
+			ctx.Fatal("Invalid Gateway ID")
+		}
+
+		act := util.GetAccount(ctx)
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_delete.go]: `act` -> `account`

File: ttnctl/cmd/gateways_edit.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+		gatewayID := args[0]
+		if !api.ValidID(gatewayID) {
+			ctx.Fatal("Invalid Gateway ID")
+		}
+
+		var edits account.GatewayEdits
+
+		frequencyPlan, err := cmd.Flags().GetString("frequency-plan")
+		if err != nil {
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_edit.go]: Also check if it's a valid frequency plan or does the server do this?

File: ttnctl/cmd/gateways_edit.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+		if locationStr != "" {
+			location, err := util.ParseLocation(locationStr)
+			if err != nil {
+				ctx.WithError(err).Fatal("Invalid location")
+			}
+			edits.Location = location
+		}
+
+		act := util.GetAccount(ctx)
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_edit.go]: `act` -> `account`

File: ttnctl/cmd/gateways_list.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/gosuri/uitable"
+	"github.com/spf13/cobra"
+)
+
+var gatewaysListCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List your gateways",
+	Long:  `ttnctl gateways list can be used to list the gateways you have access to`,
+	Run: func(cmd *cobra.Command, args []string) {
+		act := util.GetAccount(ctx)
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_list.go]: `act` -> `account`

File: ttnctl/cmd/gateways_register.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+		var err error
+		var location *account.Location
+		if len(args) == 3 {
+			location, err = util.ParseLocation(args[2])
+			if err != nil {
+				ctx.WithError(err).Fatal("Invalid location")
+			}
+		}
+
+		act := util.GetAccount(ctx)
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_register.go]: `act` -> `account`

File: ttnctl/util/location.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	"strconv"
+	"strings"
+
+	"github.com/TheThingsNetwork/ttn/core/account"
+)
+
+func ParseLocation(locationStr string) (*account.Location, error) {
+	parts := strings.Split(locationStr, ":")
+	if len(parts) != 2 {
+		return nil, errors.New("Location should be on the <lat>:<lng> format")
```
Comment (htdvisser) [inline ttnctl/util/location.go]: `<latitude>,<longitude>`

File: ttnctl/util/location.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	lat, err := strconv.ParseFloat(parts[0], 64)
+	if err != nil {
+		return nil, err
+	}
+
+	if lat < 0 || lat > 90 {
+		return nil, errors.New("Latitude should be in range [0, 90]")
```
Comment (htdvisser) [inline ttnctl/util/location.go]: Sao Paulo is `(-23.550520,-46.633309)`

File: ttnctl/util/location.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	lng, err := strconv.ParseFloat(parts[1], 64)
+	if err != nil {
+		return nil, err
+	}
+
+	if lng < 0 || lng > 180 {
+		return nil, errors.New("Longitude should be in range [0, 180]")
```
Comment (htdvisser) [inline ttnctl/util/location.go]: Sao Paulo is `(-23.550520,-46.633309)`

**Comment by romeovs** [comment]
@htdvisser Done

File: ttnctl/cmd/gateways_list.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+		gateways, err := act.ListGateways()
+		if err != nil {
+			ctx.WithError(err).Fatal("Could not list gateways")
+		}
+
+		table := uitable.New()
+		table.MaxColWidth = 70
+		table.AddRow("", "ID", "Activated", "Frequency Plan", "Lat", "Lng")
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_list.go]: `"Lat", "Lng"` -> `"Coordinates"`

File: ttnctl/cmd/gateways_list.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+		for i, gateway := range gateways {
+			var lat float64
+			var lng float64
+			if gateway.Location != nil {
+				lat = gateway.Location.Latitude
+				lng = gateway.Location.Longitude
+			}
+			table.AddRow(i+1, gateway.ID, gateway.Activated, gateway.FrequencyPlan, lat, lng)
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_list.go]: `lat, lng` -> `fmt.Sprintf("(%f,%f)", lat, lng)`

File: ttnctl/cmd/gateways_register.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/TheThingsNetwork/ttn/core/account"
+	"github.com/TheThingsNetwork/ttn/ttnctl/util"
+	"github.com/spf13/cobra"
+)
+
+var gatewaysRegisterCmd = &cobra.Command{
+	Use:   "register [GatewayID] [FrequencyPlan] [Location]",
+	Short: "Register a gateway",
+	Long:  `ttnctl gateways register can be used to register a gateway`,
+	Example: `$ ttnctl gateways register test US 52.37403:4.88968
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_register.go]: Please use coordinates in a standard format: `52.37403,4.88968`

File: ttnctl/cmd/gateways_status.go [hunk 13]
```diff
@@ -14,11 +14,11 @@ import (
...
 )
 
-var gatewayStatusCmd = &cobra.Command{
+var gatewaysStatusCmd = &cobra.Command{
 	Use:   "status [gatewayID]",
 	Short: "Get status of a gateway",
-	Long:  `ttnctl gateway status can be used to get status of gateways.`,
-	Example: `$ ttnctl gateway status eui-0000024b08060030
+	Long:  `ttnctl gateways status can be used to get status of gateways.`,
+	Example: `$ ttnctl gateways status eui-0000024b08060030
```
Comment (htdvisser) [inline ttnctl/cmd/gateways_status.go]: You're using `test` in the other examples

File: ttnctl/util/location.go [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"errors"
+	"strconv"
+	"strings"
+
+	"github.com/TheThingsNetwork/ttn/core/account"
+)
+
+func ParseLocation(locationStr string) (*account.Location, error) {
+	parts := strings.Split(locationStr, ":")
```
Comment (htdvisser) [inline ttnctl/util/location.go]: Split on `,`

File: ttnctl/cmd/gateways_delete.go (hunk 1 above)
Comment (romeovs) [inline ttnctl/cmd/gateways_delete.go]: fixed in latest commit

File: ttnctl/cmd/gateways_delete.go (hunk 2 above)
Comment (romeovs) [inline ttnctl/cmd/gateways_delete.go]: won't work because it clashes with the `account` package

File: ttnctl/cmd/gateways_edit.go (hunk 3 above)
Comment (romeovs) [inline ttnctl/cmd/gateways_edit.go]: The account server has a list of valid frequency plans and does the checking.

File: ttnctl/cmd/gateways_delete.go (hunk 2 above)
Comment (htdvisser) [inline ttnctl/cmd/gateways_delete.go]: you're not using `account` package here

**Comment by romeovs** [comment]
@htdvisser changes are implemented

**Pull request merged**